perfkitbenchmarker/linux_benchmarks/netperf_benchmark.py

```diff
@@ -93,25 +106,157 @@ def Prepare(benchmark_spec):
 
-  vms[1].RemoteCommand('%s -p %s' %
-                       (netperf.NETSERVER_PATH, COMMAND_PORT))
+  Args:
+    histogram: A dict mapping values to the number of samples with that value.
+
+  Returns:
+    A dict mapping stat names to their values.
+  """
+  stats = {}
+
+  # Histogram data in list form sorted by key
+  by_value = sorted([(value, count) for value, count in histogram.items()],
+                    key=lambda x: x[0])
+  total_count = sum(histogram.values())
+
+  cur_value_index = 0  # Current index in by_value
+  cur_index = 0  # Number of values we've passed so far
+  for p in percentiles:
+    index = int(float(total_count) * float(p) / 100.0)
+    index = min(index, total_count - 1)  # Handle 100th percentile
+    for value, count in by_value[cur_value_index:]:
+      if cur_index + count > index:
+        stats['p%s' % str(p)] = by_value[cur_value_index][0]
+        break
+      else:
+        cur_index += count
+        cur_value_index += 1
+
+  # Compute stddev
+  value_sum = float(sum([value * count for value, count in histogram.items()]))
+  average = value_sum / float(total_count)
+  if total_count > 1:
+    total_of_squares = sum([(value - average) ** 2 * count
+                            for value, count in histogram.items()])
+    stats['stddev'] = (total_of_squares / (total_count - 1)) ** 0.5
+  else:
+    stats['stddev'] = 0
+  return stats
 
 
-def RunNetperf(vm, benchmark_name, server_ip):
-  """Spawns netperf on a remove VM, parses results.
+def _ParseNetperfOutput(stdout, metadata, benchmark_name,
+                        enable_latency_histograms):
+  """Parses the stdout of a single netperf process.
+
+  Args:
+    stdout: the stdout of the netperf process
+    metadata: metadata for any sample.Sample objects we create
+
+  Returns:
+    A tuple containing (throughput_sample, latency_samples, latency_histogram)
+  """
+  # Don't modify the metadata dict that was passed in
+  metadata = metadata.copy()
+
+  fp = io.StringIO(stdout)
+  # "-o" flag above specifies CSV output, but there is one extra header line:
+  banner = next(fp)
+  assert banner.startswith('MIGRATED'), stdout
+  r = csv.DictReader(fp)
+  results = next(r)
+  logging.info('Netperf Results: %s', results)
+  assert 'Throughput' in results
+
+  # Create the throughput sample
+  throughput = float(results['Throughput'])
+  unit = {'Trans/s': TRANSACTIONS_PER_SECOND,  # *_RR
+          '10^6bits/s': MBPS}[results['Throughput Units']]  # TCP_STREAM
+  if unit == MBPS:
+    metric = '%s_Throughput' % benchmark_name
+  else:
+    metric = '%s_Transaction_Rate' % benchmark_name
+  meta_keys = [('Confidence Iterations Run', 'confidence_iter'),
+               ('Throughput Confidence Width (%)', 'confidence_width_percent')]
+  metadata.update({meta_key: results[np_key] for np_key, meta_key in meta_keys})
+  throughput_sample = sample.Sample(metric, throughput, unit, metadata)
+
+  # No tail latency for throughput.
+  if unit == MBPS:
```

Comment:
I also don't like this if-statement. It was how I found it. Now that I think about it, the logic around `enable_latency_histograms` eliminates the need for this if-statement. Removed!